### PR TITLE
Standardize on non-double-prefixed metrics

### PR DESCRIPTION
**Pull request by skpratt**
### Description
We have dueling metrics patterns - sometimes we hard-code the "consul" prefix, but then we also automatically prepend it again to all metrics. See [Issue 13494](https://github.com/hashicorp/consul/issues/13494) for more info.

This PR takes the first step toward reconciling these patterns by establishing reliance on auto-prepending prefixes as the preferred pattern, and declaring existing metrics that hard-code prefixes as deprecated in help strings.

It applies the duplication tactic in examples such as https://github.com/hashicorp/consul/blob/c04515a844cf9348198fb16c637466604a94fc09/agent/consul/fsm/commands_oss.go#L351 to all legacy metrics.

### Testing & Reproduction steps
* This does not eliminate double-prefixed metrics, but it adds a single-prefixed metric for every double-prefixed metrics, so that they can be eventually deleted.
* Tests updated to check for new metrics.

### Links
[Issue 13494](https://github.com/hashicorp/consul/issues/13494)

### PR Checklist

* [x] updated test coverage
* [N/A] external facing docs updated (this change makes metrics consistent with docs)
* [N/A] not a security concern

### CODE REVIEW

File: agent/cache/cache.go [hunk 1]
```diff
@@ -37,6 +37,10 @@ import (
 var Gauges = []prometheus.GaugeDefinition{
 	{
 		Name: []string{"consul", "cache", "entries_count"},
+		Help: "Deprecated - please use cache_entries_count instead.",
```
Comment (acpana) [inline agent/cache/cache.go:40]: Nice! I think ideally, this is all that is needed.

I think, in practice, IIRC, these messages only get "emitted" on the dummy metrics like https://github.com/hashicorp/consul/issues/13495 . Apologies if you are already well aware of this but I think for real world metrics (especially label based), the help message is not populate properly. But that's not something to be fixed in `consul`, rather something to get fixed in `go-metrics` I think.

File: agent/consul/leader_peering.go [hunk 2]
```diff
@@ -30,8 +31,13 @@ import (
 	"github.com/hashicorp/consul/proto/pbpeerstream"
 )
 
+var leaderExportedServicesCountKeyDeprecated = []string{"consul", "peering", "exported_services"}
 var leaderExportedServicesCountKey = []string{"consul", "peering", "exported_services"}
```
Comment (acpana) [inline agent/consul/leader_peering.go:35]: should line 33 be without the `consul` prefix then ?

Also 🤦🏼 since I added that code line and forgot about the double prefixing issue.

Again, not sure if this is widely known so please bear with me as I brain dump something about prefixing these w the service name, ie `consul`.

This is mostly a Promethues thing, so this func runs over all metrics (gauges, counters, summaries) and prepends the service name: https://github.com/hashicorp/consul/blob/main/agent/setup.go#L192-L347 

But that only works if they are predeclared somewhere like `LeaderPeeringMetrics` below.

So it may be possible that some of these metrics are never predeclared (not the peering metrics here but some of the older ones). In that case, emitting the `consul` prefix might not be ideal. I've not spent time looking at which metrics don't actually get predeclared so I am just offering this as a warning -- apologies in advance if it's a wild goose chase instead.

File: agent/consul/usagemetrics/usagemetrics_oss.go [hunk 3]
```diff
@@ -25,6 +30,11 @@ func (u *UsageMetricsReporter) emitPeeringUsage(peeringUsage state.PeeringUsage)
 		float32(peeringUsage.Peerings),
 		u.metricLabels,
 	)
+	metrics.SetGaugeWithLabels(
```
Comment (acpana) [inline agent/consul/usagemetrics/usagemetrics_oss.go:33]: nice! since these are all additive changes, they merge cleanly OSS-->ENT but it would be nice if we had ENT tests mirroring this. If that adds any value of course.

**Review by acpana — COMMENTED** [review]
This looks great !! 🥇 thanks for taking it on sarah 💯 

---

I left a couple of brain dumpy, non actionable comments.

Also, for this PR, did you want to create a changelog? I think it could be useful to call out these changes (deprecating the metrics, emitting 2x for deprecated metrics). Totally fine to add it as a follow up PR too but I think it'd be cleaner to have it in here.

Finally, I did a search and found a few more metrics prefixed with `"consul", "fsm"` in `commands_oss.go`. Not sure if those qualify for this patch or if we wanted to tackle those as well.

File: agent/cache/cache.go (hunk 1 above)
Comment (skpratt) [reply agent/cache/cache.go:40]: Yep. That will have to be fixed separately.

File: agent/consul/usagemetrics/usagemetrics_oss.go (hunk 3 above)
Comment (skpratt) [reply agent/consul/usagemetrics/usagemetrics_oss.go:33]: Will add in ENT project

File: agent/consul/leader_peering.go (hunk 2 above)
Comment (skpratt) [reply agent/consul/leader_peering.go:35]: thanks for that catch! it looks like all of the metrics updated in this PR are predeclared.  But enforcing that should be a follow up to this.

**Comment by skpratt** [comment]
> 

Great points! Thank you for adding them here, it will greatly help with guiding follow up work.

File: agent/cache/cache.go (hunk 1 above)
Comment (eculver) [reply agent/cache/cache.go:40]: Should we get this fixed in `go-metrics` first so that we can deliver the help message?

File: agent/consul/usagemetrics/usagemetrics_oss.go (hunk 3 above)
Comment (eculver) [reply agent/consul/usagemetrics/usagemetrics_oss.go:33]: Alternatively, if these are additive and the ENT tests aren't any different, can we just add them here?

File: agent/cache/cache.go (hunk 1 above)
Comment (skpratt) [reply agent/cache/cache.go:40]: there is a workaround for getting the help strings - as acpana pointed to in #13495 - and I think we should consider that a separate issue with bigger scope

File: agent/consul/usagemetrics/usagemetrics_oss.go (hunk 3 above)
Comment (skpratt) [reply agent/consul/usagemetrics/usagemetrics_oss.go:33]: It looks like the ENT tests will need updates, similar to the test additions here.

**Review by eculver — APPROVED** [review]
After some discussion with @skpratt offline, I think this is good to go. We discussed updating the code to add `TODO`s for the next steps but it seems like creating issues for these follow-ups is a better place to capture them. We also decided that since this whole process of getting to a world without double-prefixed metrics is going to be a long one, we might as well not make any false promises in the changelog and/or docs as to when this would actually come to fruition.

TL;DR - looks great!
